src/fsharp/TypeChecker.fs

```diff
@@ -3518,14 +3518,57 @@ let (|ExprAsPat|_|) (f: SynExpr) =
             None
     | _ -> None
 
+/// Check if a computation or sequence expression is syntactically free of 'yield' (though not yield!)
+let YieldFree cenv expr =
+    let langVersionSupportsImplicitYield = cenv.g.langVersion.SupportsFeature LanguageFeature.ImplicitYield
+    let rec YieldFree expr =
+        match expr with
+        | SynExpr.Sequential (_, _, e1, e2, _) ->
+            YieldFree e1 && YieldFree e2
+
+        | SynExpr.IfThenElse (_, e2, e3opt, _, _, _, _) ->
+            YieldFree e2 && Option.forall YieldFree e3opt
+
+        | SynExpr.TryWith (e1, _, clauses, _, _, _, _) ->
+            YieldFree e1 && clauses |> List.forall (fun (Clause(_, _, e, _, _)) -> YieldFree e)
+
+        | (SynExpr.Match (_, _, clauses, _) | SynExpr.MatchBang (_, _, clauses, _)) ->
+            clauses |> List.forall (fun (Clause(_, _, e, _, _)) -> YieldFree e)
+
+        | SynExpr.For (_, _, _, _, _, body, _)
+        | SynExpr.TryFinally (body, _, _, _, _)
+        | SynExpr.LetOrUse (_, _, _, body, _)
+        | SynExpr.While (_, _, body, _)
+        | SynExpr.ForEach (_, _, _, _, _, body, _) ->
+            YieldFree body
+
+        | SynExpr.LetOrUseBang(_, _, _, _, _, body, _) ->
+            if langVersionSupportsImplicitYield then
+                YieldFree body
+            else
+                false
+
+        | SynExpr.YieldOrReturn((true, _), _, _) when langVersionSupportsImplicitYield -> false
+
+        | SynExpr.YieldOrReturnFrom _
+        | SynExpr.YieldOrReturn _
+        | SynExpr.ImplicitZero _
+        | SynExpr.Do _ when not langVersionSupportsImplicitYield -> false
```

Comment:
Nit: could this be moved up so that it's below `YieldOrReturn`?

This was a little hard for me to read since I had was mistakenly thinking that the previous cases would also evaluate to `false` before I realized there was a guard here.

Reply:
I think the aim is to re-introduce the exact same code that was present for prior versions of F#, so we should keep it exactly as it was

Reply:
@KevinRansom Can you move this back to the matching location so the diff is minimized?

Reply:
@ dsyme, this is the original location … You moved it, once it was no longer necessary in |SimpleSemicolonSequence|.  However, it needs to be in there to verify YieldFree (ness) for IfThenElse.